java/configuration/src/test/java/sleeper/configuration/properties/InstancePropertyGroupingTest.java

```diff
@@ -0,0 +1,111 @@
+                .containsAll(SystemDefinedInstanceProperty.getAll());
+    }
+
+    @Test
+    void shouldOrderPropertiesBasedOnGroup() {
+        // Given
+        List<InstanceProperty> propertyList = new ArrayList<>();
+        InstanceProperty property1 = userProperty("user.property.1", BULK_IMPORT, propertyList::add);
+        InstanceProperty property2 = userProperty("user.property.2", INGEST, propertyList::add);
+        InstanceProperty property3 = userProperty("user.property.3", COMMON, propertyList::add);
+
+        // When
+        List<InstanceProperty> sortedPropertyList = sortPropertiesByGroup(propertyList);
+
+        // Then
+        assertThat(sortedPropertyList)
+                .containsExactly(property3, property2, property1);
+    }
+
+    @Test
+    void shouldOrderPropertiesWithDifferentTypesBasedOnGroup() {
+        // Given
+        List<InstanceProperty> propertyList = new ArrayList<>();
+        InstanceProperty userProperty1 = userProperty("user.property.1", BULK_IMPORT, propertyList::add);
+        InstanceProperty userProperty2 = userProperty("user.property.2", INGEST, propertyList::add);
+        InstanceProperty userProperty3 = userProperty("user.property.3", COMMON, propertyList::add);
+        InstanceProperty systemProperty1 = systemProperty("system.property.1", BULK_IMPORT, propertyList::add);
+        InstanceProperty systemProperty2 = systemProperty("system.property.2", INGEST, propertyList::add);
+        InstanceProperty systemProperty3 = systemProperty("system.property.3", COMMON, propertyList::add);
+
+        // When
+        List<InstanceProperty> sortedPropertyList = sortPropertiesByGroup(propertyList);
+
+        // Then
+        assertThat(sortedPropertyList)
+                .containsExactly(userProperty3, systemProperty3, userProperty2, systemProperty2, userProperty1, systemProperty1);
+    }
+
+    @Test
+    void shouldOrderPropertiesWithInterweavingTypesBasedOnGroup() {
+        // Given
+        List<InstanceProperty> propertyList = new ArrayList<>();
+        InstanceProperty userProperty1 = userProperty("user.property.1", BULK_IMPORT, propertyList::add);
+        InstanceProperty systemProperty1 = systemProperty("system.property.1", BULK_IMPORT, propertyList::add);
+        InstanceProperty userProperty2 = userProperty("user.property.2", INGEST, propertyList::add);
+        InstanceProperty systemProperty2 = systemProperty("system.property.2", INGEST, propertyList::add);
+        InstanceProperty userProperty3 = userProperty("user.property.3", COMMON, propertyList::add);
+        InstanceProperty systemProperty3 = systemProperty("system.property.3", COMMON, propertyList::add);
+
+        // When
+        List<InstanceProperty> sortedPropertyList = sortPropertiesByGroup(propertyList);
+
+        // Then
+        assertThat(sortedPropertyList)
+                .containsExactly(userProperty3, systemProperty3, userProperty2, systemProperty2, userProperty1, systemProperty1);
+    }
```

Comment:
These last two tests make it sound like the important thing is the types, when the implementation doesn't actually care what types there are. In practice the properties will be passed in as a block of one type then a block of the other, but currently what's tested doesn't know or care about the types.

The behaviour we want is, given we have two blocks of properties, where properties of each group are together within each block (though not all tests need to care about the blocks):

- The properties are reordered so that the groups are together.
- The order that the properties were in in their blocks is preserved after the reordering.
- The order of the blocks is preserved after the reordering.